src/engine/Metavariable_pattern.ml

```diff
@@ -118,53 +118,79 @@ let satisfies_metavar_pattern_condition nested_formula_has_matches env r mvar
-                (Xlang.to_string xlang);
-              (* We re-parse the matched text as `xlang`. *)
-              Xpattern_matcher.with_tmp_file ~str:content ~ext:"mvar-pattern"
-                (fun file ->
-                  let lazy_ast_and_errors =
-                    lazy
-                      (match xlang with
-                      | L (lang, _) ->
-                          let { Parse_target.ast; skipped_tokens; _ } =
-                            Parse_target.parse_and_resolve_name lang file
-                          in
-                          (* TODO: If we wanted to report the parse errors
-                           * then we should fix the parse info with
-                           * Parse_info.adjust_info_wrt_base! *)
-                          if skipped_tokens <> [] then
-                            pr2
-                              (spf
-                                 "rule %s: metavariable-pattern: failed to \
-                                  fully parse the content of %s"
-                                 (fst env.rule.Rule.id) mvar);
-                          (ast, skipped_tokens)
-                      | LRegex
-                      | LGeneric ->
-                          failwith "requesting generic AST for LRegex|LGeneric")
-                  in
-                  let xtarget =
-                    {
-                      Xtarget.file;
-                      xlang;
-                      lazy_ast_and_errors;
-                      lazy_content = lazy content;
-                    }
-                  in
-                  nested_formula_has_matches { env with xtarget } formula
-                    (Some r'))
-          | Some _lang, mval ->
-              (* This is not necessarily an error in the rule, e.g. you may be
-               * matching `$STRING + ...` and then add a metavariable-pattern on
-               * `$STRING`. This will only work when `$STRING` binds to some text
-               * but it can naturally bind to other string expressions. *)
-              logger#debug
-                "metavariable-pattern failed because the content of %s is not \
-                 text: %s"
-                mvar (MV.show_mvalue mval);
-              false))
+          | Some xlang -> (
+              let content =
+                (* Previously we only allowed metavariable-pattern with a
+                 * different language when the metavariable was bound to a
+                 * string. For these cases, we keep the previous behavior, which
+                 * I (nmote) believe makes only the contents of the string
+                 * available for matching in the sub language, and not the
+                 * enclosing quotes.
+                 *
+                 * However, now we also allow non-text metavariable bindings to
+                 * match in generic mode. Because non-text metavariable bindings
+                 * are highly unlikely to be syntactically valid in any
```

Comment:
Do tests break if you allow this for any language and not just `generic` ?

Reply:
No. The one test that broke was updated in https://github.com/returntocorp/semgrep-rules/pull/2691. @LewisArdern agreed that the new behavior here makes sense, and updated the rule accordingly.

I wouldn't object strongly to allowing this for any language, but I would want to see a motivating use case.